File: claude-code/extractors_swift/Sources/Extractors/HtmlExtractor.swift
import SwiftSoup

/// Extracts structured content from HTML using SwiftSoup.
public struct HtmlExtractor {

    public init() {}

    /// Result of HTML extraction.
    public struct ExtractionResult {
        public let title: String
        public let content: String
        public let headings: [Heading]
        public let codeBlocks: [CodeBlock]
        public let links: [PageLink]
    }

    public struct Heading {
        public let level: Int
        public let text: String
    }

    public struct CodeBlock {
        public let language: String
        public let content: String
        public let lineCount: Int
    }

    public struct PageLink {
        public let text: String
        public let href: String
        public let isInternal: Bool
    }

    private static let navSelectors = [
        "nav", "header", "footer",
        "[role=navigation]", ".nav", ".sidebar", ".menu", ".breadcrumb"
    ]

    private static let contentSelectors = [
        "main", "article", "[role=main]", "#content", "#main-content",
        ".content", ".main-content", ".post-content", ".markdown-body", ".docs-content"
    ]

    /// Extract structured content from HTML.
    public func extract(html: String) throws -> ExtractionResult {
        let doc = try SwiftSoup.parse(html)

        // Remove non-content elements
        try doc.select("script, style, noscript, iframe, svg").remove()
        for selector in Self.navSelectors {
            try doc.select(selector).remove()
        }

        let mainContent = try findMainContent(doc: doc)
        let title = try extractTitle(doc: doc)
        let content = try elementToMarkdown(element: mainContent)
        let headings = try extractHeadings(element: mainContent)
        let codeBlocks = try extractCodeBlocks(element: mainContent)
        let links = try extractLinks(element: mainContent)

        return ExtractionResult(
            title: title,
            content: content.trimmingCharacters(in: .whitespacesAndNewlines),
            headings: headings,
            codeBlocks: codeBlocks,
            links: links
        )
    }

    private func extractTitle(doc: Document) throws -> String {
        if let ogTitle = try doc.select("meta[property=og:title]").first() {
            let content = try ogTitle.attr("content").trimmingCharacters(in: .whitespaces)
            if !content.isEmpty { return content }
        }

        if let titleEl = try doc.select("title").first() {
            let text = try titleEl.text().trimmingCharacters(in: .whitespaces)
            if !text.isEmpty { return text }
        }

        if let h1 = try doc.select("h1").first() {
            return try h1.text().trimmingCharacters(in: .whitespaces)
        }

        return ""
    }

    private func findMainContent(doc: Document) throws -> Element {
        for selector in Self.contentSelectors {
            if let el = try doc.select(selector).first() {
                let text = try el.text().trimmingCharacters(in: .whitespaces)
                if text.count > 100 {
                    return el
                }
            }
        }

        if let body = doc.body() {
            return body
        }

        return doc
    }

    private func elementToMarkdown(element: Element) throws -> String {
        var output = ""

        for child in try element.select("h1, h2, h3, h4, h5, h6, p, pre, li") {
            let tagName = child.tagName().lowercased()
            let text = try child.text().trimmingCharacters(in: .whitespaces)

            switch tagName {
            case "h1": output += "\n\n# \(text)\n\n"
            case "h2": output += "\n\n## \(text)\n\n"
            case "h3": output += "\n\n### \(text)\n\n"
            case "h4": output += "\n\n#### \(text)\n\n"
            case "h5": output += "\n\n##### \(text)\n\n"
            case "h6": output += "\n\n###### \(text)\n\n"
            case "p":
                if !text.isEmpty {
                    output += "\n\n\(text)\n\n"
                }
            case "pre":
                let lang = try getCodeLanguage(preElement: child)
                output += "\n\n```\(lang)\n\(text)\n```\n\n"
            case "li":
                output += "\n- \(text)"
            default:
                break
            }
        }

        return output
    }

    private func extractHeadings(element: Element) throws -> [Heading] {
        var headings: [Heading] = []

        for h in try element.select("h1, h2, h3, h4, h5, h6") {
            let tagName = h.tagName()
            let level = Int(String(tagName.last ?? "0")) ?? 0
            let text = try h.text().trimmingCharacters(in: .whitespaces)
            headings.append(Heading(level: level, text: text))
        }

        return headings
    }

    private func extractCodeBlocks(element: Element) throws -> [CodeBlock] {
        var blocks: [CodeBlock] = []

        for pre in try element.select("pre") {
            let content = try pre.text()
            let language = try getCodeLanguage(preElement: pre)
            let lineCount = content.split(separator: "\n").count

            blocks.append(CodeBlock(
                language: language,
                content: content,
                lineCount: lineCount
            ))
        }

        return blocks
    }

    private func extractLinks(element: Element) throws -> [PageLink] {
        var links: [PageLink] = []

        for a in try element.select("a[href]") {
            let href = try a.attr("href")
            if href.hasPrefix("#") || href.hasPrefix("javascript:") { continue }

            let text = try a.text().trimmingCharacters(in: .whitespaces)
            let isInternal = href.hasPrefix("/") || href.hasPrefix("./") || href.hasPrefix("../")

            links.append(PageLink(text: text, href: href, isInternal: isInternal))
        }

        return links
    }

    private func getCodeLanguage(preElement: Element) throws -> String {
        if let code = try preElement.select("code").first() {
            let classes = try code.className().split(separator: " ")
            for cls in classes {
                let clsStr = String(cls)
                if clsStr.hasPrefix("language-") { return String(clsStr.dropFirst(9)) }
                if clsStr.hasPrefix("lang-") { return String(clsStr.dropFirst(5)) }
                if clsStr.hasPrefix("highlight-") { return String(clsStr.dropFirst(10)) }
            }

            let dataLang = try code.attr("data-lang")
            if !dataLang.isEmpty { return dataLang }

            let dataLanguage = try code.attr("data-language")
            if !dataLanguage.isEmpty { return dataLanguage }
        }

        return ""
    }
}

File: claude-code/extractors_java/src/main/java/com/researchers/HtmlExtractor.java
package com.researchers;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Extracts structured content from HTML using Jsoup.
 */
public class HtmlExtractor {

    private static final String[] NAV_SELECTORS = {
            "nav", "header", "footer",
            "[role=navigation]", ".nav", ".sidebar", ".menu", ".breadcrumb"
    };

    private static final String[] CONTENT_SELECTORS = {
            "main", "article", "[role=main]", "#content", "#main-content",
            ".content", ".main-content", ".post-content", ".markdown-body", ".docs-content"
    };

    /**
     * Extract structured content from HTML.
     *
     * @param html Raw HTML string
     * @return Map containing title, content (as markdown), headings, code blocks, and links
     */
    public Map<String, Object> extract(String html) {
        Document doc = Jsoup.parse(html);

        // Remove non-content elements
        doc.select("script, style, noscript, iframe, svg").remove();
        for (String selector : NAV_SELECTORS) {
            doc.select(selector).remove();
        }

        Element mainContent = findMainContent(doc);

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("title", extractTitle(doc));
        result.put("content", elementToMarkdown(mainContent));
        result.put("headings", extractHeadings(mainContent));
        result.put("codeBlocks", extractCodeBlocks(mainContent));
        result.put("links", extractLinks(mainContent));

        return result;
    }

    private String extractTitle(Document doc) {
        Element ogTitle = doc.selectFirst("meta[property=og:title]");
        if (ogTitle != null) {
            String content = ogTitle.attr("content");
            if (!content.isEmpty()) return content;
        }

        Element title = doc.selectFirst("title");
        if (title != null) {
            String text = title.text().trim();
            if (!text.isEmpty()) return text;
        }

        Element h1 = doc.selectFirst("h1");
        if (h1 != null) return h1.text().trim();

        return "";
    }

    private Element findMainContent(Document doc) {
        for (String selector : CONTENT_SELECTORS) {
            Element el = doc.selectFirst(selector);
            if (el != null && el.text().trim().length() > 100) {
                return el;
            }
        }

        Element body = doc.body();
        return body != null ? body : doc;
    }

    private String elementToMarkdown(Element element) {
        StringBuilder sb = new StringBuilder();

        for (Element child : element.getAllElements()) {
            String tagName = child.tagName().toLowerCase();

            switch (tagName) {
                case "h1":
                    sb.append("\n\n# ").append(child.ownText()).append("\n\n");
                    break;
                case "h2":
                    sb.append("\n\n## ").append(child.ownText()).append("\n\n");
                    break;
                case "h3":
                    sb.append("\n\n### ").append(child.ownText()).append("\n\n");
                    break;
                case "h4":
                    sb.append("\n\n#### ").append(child.ownText()).append("\n\n");
                    break;
                case "h5":
                    sb.append("\n\n##### ").append(child.ownText()).append("\n\n");
                    break;
                case "h6":
                    sb.append("\n\n###### ").append(child.ownText()).append("\n\n");
                    break;
                case "p":
                    String text = child.text().trim();
                    if (!text.isEmpty()) {
                        sb.append("\n\n").append(text).append("\n\n");
                    }
                    break;
                case "pre":
                    String lang = getCodeLanguage(child);
                    String code = child.text();
                    sb.append("\n\n```").append(lang).append("\n").append(code).append("\n```\n\n");
                    break;
                case "li":
                    sb.append("\n- ").append(child.ownText());
                    break;
                default:
                    break;
            }
        }

        return cleanMarkdown(sb.toString());
    }

    private List<Map<String, String>> extractHeadings(Element element) {
        List<Map<String, String>> headings = new ArrayList<>();

        for (Element heading : element.select("h1, h2, h3, h4, h5, h6")) {
            Map<String, String> h = new LinkedHashMap<>();
            h.put("level", heading.tagName().substring(1));
            h.put("text", heading.text().trim());
            h.put("id", heading.id());
            headings.add(h);
        }

        return headings;
    }

    private List<Map<String, String>> extractCodeBlocks(Element element) {
        List<Map<String, String>> blocks = new ArrayList<>();

        for (Element pre : element.select("pre")) {
            Map<String, String> block = new LinkedHashMap<>();
            block.put("language", getCodeLanguage(pre));
            block.put("content", pre.text());
            block.put("lineCount", String.valueOf(pre.text().split("\n").length));
            blocks.add(block);
        }

        return blocks;
    }

    private List<Map<String, String>> extractLinks(Element element) {
        List<Map<String, String>> links = new ArrayList<>();

        for (Element a : element.select("a[href]")) {
            String href = a.attr("href");
            if (href.startsWith("#") || href.startsWith("javascript:")) continue;

            Map<String, String> link = new LinkedHashMap<>();
            link.put("text", a.text().trim());
            link.put("href", href);
            link.put("isInternal", String.valueOf(
                    href.startsWith("/") || href.startsWith("./") || href.startsWith("../")
            ));
            links.add(link);
        }

        return links;
    }

    private String getCodeLanguage(Element preElement) {
        Element code = preElement.selectFirst("code");
        if (code != null) {
            for (String cls : code.classNames()) {
                if (cls.startsWith("language-")) return cls.substring(9);
                if (cls.startsWith("lang-")) return cls.substring(5);
                if (cls.startsWith("highlight-")) return cls.substring(10);
            }
            String dataLang = code.attr("data-lang");
            if (!dataLang.isEmpty()) return dataLang;
            String dataLanguage = code.attr("data-language");
            if (!dataLanguage.isEmpty()) return dataLanguage;
        }
        return "";
    }

    private String cleanMarkdown(String md) {
        // Remove excessive blank lines
        md = md.replaceAll("\n{4,}", "\n\n\n");

        // Trim trailing whitespace from lines
        StringBuilder sb = new StringBuilder();
        for (String line : md.split("\n")) {
            sb.append(line.stripTrailing()).append("\n");
        }

        return sb.toString().trim();
    }
}
